hail/python/hail/vds/sample_qc.py

```diff
@@ -0,0 +1,310 @@
+        bases_over_gq_threshold=hl.tuple(
+            x + y for x, y in zip(vmt_sample_qc.gq_dp_exprs.gq, rmt_sample_qc.bases_over_gq_threshold)
+        ),
+        **joined_dp_expr,
+    )
+
+
+@typecheck(vds=VariantDataset, gq_bins=sequenceof(int), dp_bins=sequenceof(int), dp_field=nullable(str))
+def sample_qc(
+    vds: 'VariantDataset',
+    *,
+    gq_bins: 'Sequence[int]' = (0, 20, 60),
+    dp_bins: 'Sequence[int]' = (0, 1, 10, 20, 30),
+    dp_field=None,
+) -> 'Table':
+    """Compute sample quality metrics about a :class:`.VariantDataset`.
+
+    If the `dp_field` parameter is not specified, the ``DP`` is used for depth
+    if present. If no ``DP`` field is present, the ``MIN_DP`` field is used. If no ``DP``
+    or ``MIN_DP`` field is present, no depth statistics will be calculated.
+
+    Parameters
+    ----------
+    vds : :class:`.VariantDataset`
+        Dataset in VariantDataset representation.
+    name : :obj:`str`
+        Name for resulting field.
+    gq_bins : :class:`tuple` of :obj:`int`
+        Tuple containing cutoffs for genotype quality (GQ) scores.
+    dp_bins : :class:`tuple` of :obj:`int`
+        Tuple containing cutoffs for depth (DP) scores.
+    dp_field : :obj:`str`
+        Name of depth field. If not supplied, DP or MIN_DP will be used, in that order.
+
+    Returns
+    -------
+    :class:`.Table`
+        Hail Table of results, keyed by sample.
+    """
+
+    require_first_key_field_locus(vds.reference_data, 'sample_qc')
+    require_first_key_field_locus(vds.variant_data, 'sample_qc')
+
+    if 'DP' in vds.reference_data.entry:
+        ref_dp_field_to_use = 'DP'
+    elif 'MIN_DP' in vds.reference_data.entry:
+        ref_dp_field_to_use = 'MIN_DP'
+    else:
+        ref_dp_field_to_use = dp_field
+
+    vmt = vds.variant_data
+    if 'GT' not in vmt.entry:
+        vmt = vmt.annotate_entries(GT=hl.vds.lgt_to_gt(vmt.LGT, vmt.LA))
+    ac_and_atypes = vmt_sample_qc_variant_annotations(global_gt=vmt.GT, alleles=vmt.alleles)
+    variant_ac = Env.get_uid()
+    variant_atypes = Env.get_uid()
+    vmt = vmt.annotate_rows(**{variant_ac: ac_and_atypes.ac, variant_atypes: ac_and_atypes.atypes})
+    vmt_dp = vmt['DP'] if ref_dp_field_to_use is not None and 'DP' in vmt.entry else None
+    variant_results = vmt.select_cols(
+        **vmt_sample_qc(
+            global_gt=vmt.GT,
+            gq=vmt.GQ,
+            variant_ac=vmt[variant_ac],
+            variant_atypes=vmt[variant_atypes],
+            dp=vmt_dp,
+            gq_bins=gq_bins,
+            dp_bins=dp_bins,
+        )
+    ).cols()
+
+    rmt = vds.reference_data
+    rmt_dp = rmt[ref_dp_field_to_use] if ref_dp_field_to_use is not None else None
+    reference_results = rmt.select_cols(
+        **rmt_sample_qc(
+            locus=rmt.locus,
+            gq=rmt.GQ,
+            end=rmt.END,
+            dp=rmt_dp,
+            gq_bins=gq_bins,
+            dp_bins=dp_bins,
+        )
+    ).cols()
+
+    # TODO factor this out?
```

Comment:
I think it makes sense to factor out the combine variant and ref qc step. I'd expect anybody using the other factored out methods will need to do this.